src/main/kotlin/net/leanix/githubagent/services/GitHubAuthenticationService.kt

```diff
@@ -49,32 +48,52 @@ class GitHubAuthenticationService(
     fun generateAndCacheJwtToken() {
         runCatching {
             logger.info("Generating JWT token")
+            val privateKey = loadPrivateKey()
+            val jwt = createJwtToken(privateKey)
+            verifyAndCacheJwtToken(jwt)
+        }.onFailure {
+            logger.error("Failed to generate a valid jwt token", it)
+            throw it
+        }
+    }
+
+    private fun loadPrivateKey(): PrivateKey {
+        return runCatching {
             Security.addProvider(BouncyCastleProvider())
             val rsaPrivateKey: String = readPrivateKey()
             val keySpec = PKCS8EncodedKeySpec(Base64.getDecoder().decode(rsaPrivateKey))
-            val privateKey = KeyFactory.getInstance("RSA").generatePrivate(keySpec)
-            val jwt = createJwtToken(privateKey)
-            gitHubEnterpriseService.verifyJwt(jwt.getOrThrow())
-            cachingService.set("jwtToken", jwt.getOrThrow(), JWT_EXPIRATION_DURATION)
-        }.onFailure {
-            if (it is InvalidKeySpecException) {
-                throw IllegalArgumentException("The provided private key is not in a valid PKCS8 format.", it)
-            } else {
-                throw it
-            }
+            KeyFactory.getInstance("RSA").generatePrivate(keySpec)
+        }.getOrElse {
+            logger.error("Failed to load private key", it)
+            throw IllegalArgumentException(
+                "Failed to load private key, " +
+                    "the provided private key is not a valid PKCS8 key."
+            )
         }
     }
 
-    private fun createJwtToken(privateKey: PrivateKey): Result<String> {
+    private fun createJwtToken(privateKey: PrivateKey): String {
         return runCatching {
             Jwts.builder()
                 .setIssuedAt(Date())
                 .setExpiration(Date(System.currentTimeMillis() + JWT_EXPIRATION_DURATION))
                 .setIssuer(cachingService.get("githubAppId").toString())
                 .signWith(privateKey, SignatureAlgorithm.RS256)
                 .compact()
+        }.getOrElse {
+            logger.error("Failed to generate a JWT token", it)
+            throw FailedToCreateJWTException("Failed to generate a JWT token")
+        }
+    }
+
+    private fun verifyAndCacheJwtToken(jwt: String) {
+        runCatching {
+            gitHubEnterpriseService.verifyJwt(jwt)
+            cachingService.set("jwtToken", jwt, JWT_EXPIRATION_DURATION)
+            logger.info("JWT token generated and cached successfully")
         }.onFailure {
-            throw FailedToCreateJWTException("Failed to generate a valid JWT token")
+            logger.error("Failed to verify and cache JWT token", it)
+            throw it
```

Comment:
why are we logging the exception and throwing it again.
Why not log the exception where it is finally caught?

Reply:
I do get it. But I just want to make sure that while doing it we are not logging the same exception twice

Reply:
If that is not the case then I am not good with it